File: problem_3/pro3.js
/** 
   Project Euler

   Largest prime factor

   Problem 3

   The prime factors of 13195 are 5, 7, 13, and 29.

   What is the largest prime factor of the number 600851475143 ? 
*/

//solution 1 rely on smallest prime factor
function smallest_factor(n) {

         i = 2

         while(i <= Math.sqrt(n) ) { 

          if( 0 == n%i ){

             return i
          }

          i++

         }

         return n
}

function largest_factor(n) {

         m = n

         l = 1

         while( m != 1 ) {

               k = smallest_factor(m)  

               m = m / k

               l = k              
         }

  return l
}

function largest_prime_factor(N) { 

         var i = 2, 

             factors = [],

             k = -1; 
    
         while((i*i) <= N) { 
       
               if(0 == N%i) { 
 
                  factors[ ++k ] = i 

                  N /= i 

               } else { 

                  i++ 
               } 
         } 

         if( N != 1) factors[ ++k ] = N 

   return factors[ factors.length - 1 ] 
}

File: problem_3/pro3.py
# Project Euler
# Largest prime factor
# Problem 3
# The prime factors of 13195 are 5, 7, 13, and 29.
# What is the largest prime factor of the number 600851475143 ? 

#solution 1
def smallest_factor(n):
    i = 2
    while i < n:
        if 0 == n % i:
           return i
        i = i + 1
    return n

def largest_prime_factor(n):

    m = n
    l = 1
  
    while m != 1:
          if m <= 2:
             return m
          else:
             j = smallest_factor(m)
             m = m / j
             l = j
    return l

#solution 2
def largestPrimeFactor( n ):

    i, factors = 2, []
    while i*i <= n:
          if 0 == n%i:
             factors.append( i )
             n = n / i 
          else:
             i = i + 1    

    if n!= 1: factors.append(n) 

    return factors[ len( factors ) - 1 ]

#solution 1
print "Solution 1 -> ", largest_prime_factor( 600851475143 )

#solution 2
print "Solution 2 -> ", largestPrimeFactor( 600851475143 )

